EN-27 Create new projects_topics functionality

jinjagit opened the pull request:
## Checklist

- [] Added entry to changelog
Not yet - this feature is not completed by this PR.

- [x] Tests

- [x] Prepared branch for code review

## Links

- [Homepage Filters Epic](https://citizenlab.atlassian.net/browse/EN-1757)

## How urgent is a code review?

A.S.A.P - FE needs this to implement adding / removing topics (tags) of Project

jinjagit committed:
Basic migration and model + model associations

jinjagit committed:
revert unintended change to schema.rb

jinjagit committed:
Enable projects_topics creation on Projects POST request

Also return associated topics and projects_topics in response

jinjagit committed:
remove unintended duplication

jinjagit committed:
enable update of projects_topics when update project

jinjagit committed:
add tests of associated topics to projects_spec

alexander-cit commented on back/app/models/topic.rb:
```diff
@@ -22,6 +22,8 @@ def self.codes
   acts_as_list column: :ordering, top_of_list: 0, add_new_at: :top
 
+  has_many :projects_topics, dependent: :destroy
+  has_many :projects, through: :projects_topics
```
Now we have two `has_many :projects` in this model.

alexander-cit commented on back/spec/acceptance/projects_spec.rb:
```diff
@@ -534,6 +539,11 @@
 
+      example 'Clear all topics', document: false do
+        do_request(project: { topic_ids: [] })
+        expect(json_response.dig(:data, :relationships, :topics, :data).size).to eq 0
```
Isn't it 0 without `do_request(project: { topic_ids: [] })` too?

alexander-cit replied:
I still don't understand how topics are tied to projects in this test (project is created without any topics here). And so, why they're removed. Do you know how?

alexander-cit commented on back/app/models/projects_topic.rb:
```diff
@@ -0,0 +1,24 @@
+  belongs_to :project
+  belongs_to :topic
+
+  validates :project, :topic, presence: true
```
Presence is validated by default for `belongs_to`, so you can remove this line.

alexander-cit commented on back/db/migrate/20220207103216_create_projects_topics.rb:
```diff
@@ -0,0 +1,14 @@
+    create_table :projects_topics, id: :uuid do |t|
+      t.uuid :project_id
+      t.uuid :topic_id
+      t.integer :ordering
```
Is `ordering` used now? Will it be used?

alexander-cit commented on back/db/migrate/20220207103216_create_projects_topics.rb:
```diff
@@ -0,0 +1,14 @@
+class CreateProjectsTopics < ActiveRecord::Migration[6.1]
+  def change
+    create_table :projects_topics, id: :uuid do |t|
+      t.uuid :project_id
```
We should better add `null: false` and foreign key constraints for these two ids.

I would also add `null: false` to ordering and timestamps.

alexander-cit replied:
The easiest way would be to use `references` https://apidock.com/rails/ActiveRecord/ConnectionAdapters/Table/references
with `foreign_key: true, index: true`.
You won't need t.index in this case.

jinjagit commented:
@alexander-cit For some reason I can't comment on your comments...

 Yes, there are 2 `has_many :projects` relationships, but with different `through` associations. How better to do this?  

"Isn't it 0 without do_request(project: { topic_ids: [] }) too?" It would be zero if you made the request without specifying `topics_ids`, I guess, but you need the request to update, to remove the 2 existing relations, no?

"Presence is validated by default for belongs_to, so you can remove this line." Which line? You highlighted 4 lines.

"Is ordering used now? Will it be used?" I don't think we will need it in this epic, but wasn't sure so I just copied what we have for `projects_allowed_input_topics` (for now)

"We should better add null: false and foreign key constraints for these two ids. I would also add null: false to ordering and timestamps." - I thought about this, but again, copying `projects_allowed_input_topics` I thought maybe there was a reason we didn't do it there (historical reason, I mean).

alexander-cit commented:
@jinjagit
 
> Yes, there are 2 has_many :projects relationships, but with different through associations. How better to do this?

It seems you don't need `has_many :projects`. Do you?

> I guess, but you need the request to update, to remove the 2 existing relations, no?

I don't see where the relation is created, so it's better to check commenting  `do_request(project: { topic_ids: [] })` and see if the test still passes.

> Which line?

The one with validation as it's excessive `  validates :project, :topic, presence: true`

> I thought about this, but again, copying projects_allowed_input_topics I thought maybe there was a reason we didn't do it there (historical reason, I mean).

Let me know if you find any reason for it (I don't see any), but I believe by default we should be as strict as possible.

jinjagit committed:
improve migration & simplify models

jinjagit commented:
@alexander-cit  I think I have made all the changes / improvements you suggested thus far.

The test you mentioned does indeed fail if I comment out the request, with
```
Projects when admin PATCH web_api/v1/projects/:id Clear all topics
     Failure/Error: let(:json_response) { json_parse(response_body) }
     
     Rack::Test::Error:
       No response yet. Request a page first.
```
...since there is no response body due to there being no request. I suspect, however, this is not what you meant.

Note that `do_request(project: { })` also fails with
```
Failure/Error: expect(json_response.dig(:data, :relationships, :topics, :data).size).to eq 0
     
       expected: 0
            got: 2
```
...as expected / desired.

alexander-cit commented on back/app/models/topic.rb:
```diff
@@ -22,6 +22,8 @@ def self.codes
   acts_as_list column: :ordering, top_of_list: 0, add_new_at: :top
 
+  #has_many :projects_topics, dependent: :destroy
+  #has_many :projects, through: :projects_topics
```
Maybe it's better to delete these comments.

alexander-cit commented on back/app/serializers/web_api/v1/project_serializer.rb:
```diff
@@ -94,6 +94,8 @@ class WebApi::V1::ProjectSerializer < WebApi::V1::BaseSerializer
   has_many :project_images, serializer: WebApi::V1::ImageSerializer
   has_many :areas
+  has_many :topics, serializer: WebApi::V1::TopicSerializer
+  has_many :projects_topics
```
Do we need to send `projects_topics` to the client? I'm wondering how client can use it.

jinjagit committed:
remove ordering from migration

jinjagit committed:
remove comments

alexander-cit commented on back/db/migrate/20220207103216_create_projects_topics.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateProjectsTopics < ActiveRecord::Migration[6.1]
+  def change
+    create_table :projects_topics, id: :uuid do |t|
+      t.uuid :project_id, null: false, foreign_key: true, index: true
```
According to `schema.rb`, foreign key is not created. I believe you should use `references` method from the link I've sent before.

jinjagit commented:
@alexander-cit You mean like this (without `foreign_key`)?
```
class CreateProjectsTopics < ActiveRecord::Migration[6.1]
  def change
    create_table :projects_topics, id: :uuid do |t|
      t.references(:topic, null: false, index: true)
      t.references(:project, null: false, index: true)

      t.timestamps
    end
  end
end
```

alexander-cit commented:
Rather like that

```ruby
class CreateProjectsTopics < ActiveRecord::Migration[6.1]
  def change
    create_table :projects_topics, id: :uuid do |t|
      t.references(:topic, null: false, index: true, foreign_key: true)
      t.references(:project, null: false, index: true, foreign_key: true)

      t.timestamps
    end
  end
end
```

And then check the changes in schema.rb

jinjagit committed:
use t.references to define foreign_key

jinjagit commented:
@alexander-cit "Do we need to send projects_topics to the client?" - No, I don't think we do. I was wondering the same. I will remove this from the serializer. (We can always replace it, if we later discover it is useful)

jinjagit committed:
remove projects_topics from projects serializer

jinjagit committed:
Improve specs for PATCH project topics, & areas

jinjagit commented:
@alexander-cit I think the only outstanding question I have is should we keep the `schema.rb` changes that appear every time we migrate?
```
 t.index ["item_type", "item_id"], name: "index_activities_on_item"
```
becomes...
```
t.index ["item_type", "item_id"], name: "index_activities_on_item_type_and_item_id"
```

alexander-cit commented:
It seems it depends on whether DB was migrated from scratch or created by schema:dump (db:reset). We should definitely investigate and fix it, but for now it doesn't matter. You can checkout or commit this change.

alexander-cit committed:
[EN-1870] Remove unused topics/allowed_input_topics code

alexander-cit committed:
[EN-1870] Filter projects by topics (existing filter wasn't used)

alexander-cit committed:
[EN-1870] Simplify topics sorting. Get rid of unused project_id param

alexander-cit reviewed (approved):
Clean and simple :+1: 

But please do not merge. I will merge it myself when filtering is finished.

alexander-cit committed:
[EN-1870] Remove remove_childless_parents param as it's always false

alexander-cit committed:
[EN-1870] Fetch topics for filtering

luucvanderzee committed:
[EN-657] Merge branch 'EN-1870-filter-projects-by-topics' into EN-657-add-tag-filter

luucvanderzee committed:
[EN-657] Clean up apiEndpoint variables in projectTopics service

luucvanderzee committed:
[EN-657] Rename: projectTopics service -> projectAllowedInputTopics

luucvanderzee committed:
[EN-657] services/topics: move add, update and delete topic into custom_topics module

luucvanderzee committed:
[EN-657] Add typing to addTopic and updateTopic

luucvanderzee committed:
[EN-657] Rename: topics tab in project edit page -> allowed input topics

luucvanderzee committed:
[EN-657] Renaming stuff in custom_topics module (WIP)

luucvanderzee committed:
[EN-657] Refactor: rename useProjectTopics -> useProjectAllowedInputTopics and use in SortableProjectTopicList

alexander-cit committed:
[EN-1870] Send all topics-related data in projects#show

We send all the topics-related data in projects#show. This way we:
1. cache topics clearly and easily
2. do not duplicate topics in different requests
3. send just one request instead of 2
https://citizenlabco.slack.com/archives/C02K46G9SJU/p1644852760466379?thread_ts=1644842406.048499&cid=C02K46G9SJU

luucvanderzee committed:
[EN-657] Refactor TopicFilterDropdown to use new useProjectAllowedInputTopics hook

luucvanderzee committed:
[EN-657] Refactor useInputSchema to use new useProjectAllowedInputTopics hook

luucvanderzee committed:
[EN-657] Refactor ProjectTopicSelector to use new useProjectAllowedInputTopics hook

luucvanderzee committed:
[EN-657] Remove projectId param from useTopics

luucvanderzee committed:
[EN-657] Add useTopic hook

luucvanderzee committed:
[EN-657] Refactor TopicSettings/Edit/index to functional component

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
[EN-657] Delete GetTopic

luucvanderzee committed:
[EN-657] Refactor useProjectAllowedInputTopics for reusability in corresponding resource

luucvanderzee committed:
[EN-657] Make projectId in useProjectAllowedInputTopics optional

luucvanderzee committed:
[EN-657] Create GetProjectAllowedInputTopics

luucvanderzee committed:
[EN-657] Add getNestedTopicData to GetProjectAllowedInputTopics

luucvanderzee committed:
[EN-657] Replace GetTopics in PostManager by GetProjectAllowedInputTopics

luucvanderzee committed:
[EN-657] Replace GetTopics in IdeaForm and update terminology

luucvanderzee committed:
[EN-657] Remove projectId from GetTopics

alexander-cit committed:
[EN-1870][CL2-6837] Check areas/topics param removing empty folders

alexander-cit committed:
[EN-1870] Remove outdated specs (see prev commit)

alexander-cit committed:
[EN-1870][CL2-6837] Fix specs for non-EE deployment

luucvanderzee committed:
[EN-657] Fix bugs in GetProjectAllowedInputTopics

luucvanderzee committed:
[EN-657] Merge branch 'EN-657-add-tag-filter' of github.com:CitizenLabDotCo/citizenlab into EN-657-add-tag-filter

luucvanderzee committed:
[EN-657] Fix hook order issue in TopicFilterDropdown

luucvanderzee committed:
[EN-657] Merge branch 'EN-1870-filter-projects-by-topics' into EN-657-add-tag-filter

luucvanderzee committed:
[EN-657] Add new topic relationships to IProjectData

alexander-cit committed:
[EN-1870] Get back projects_allowed_input_topics#show

Fow not it seems this endpoint won't be used from FE because
projects_allowed_input_topics will always be cached in `included`
part of projects#show response.
But we add it because `included` should not be used directly,
it just caches data for future responses
(projects_allowed_input_topics#show in this case).

This commit partially reverts 39122df9c1 and 6bf0b77d87

luucvanderzee committed:
Merge branch 'EN-1870-filter-projects-by-topics' into EN-657-add-tag-filter

luucvanderzee committed:
Add projectAllowedInputTopic service

luucvanderzee committed:
Remove projectAllowedInputTopicsStream

luucvanderzee committed:
_

luucvanderzee committed:
Change how useProjectAllowedInputTopics works

luucvanderzee committed:
Change how GetProjectAllowedInputTopics works

luucvanderzee committed:
_

luucvanderzee committed:
usePages and useProjectAllowedInputTopics: use reduceErrors utility

luucvanderzee committed:
Use reduceErrors in useTopics

luucvanderzee committed:
Remove unnecessary error checking wherever useTopics is used

luucvanderzee committed:
Use reduceErrors in GetTopics

luucvanderzee committed:
Update places where useProjectAllowedInputTopics is used

luucvanderzee committed:
Update places where GetProjectAllowedInputTopics is used

luucvanderzee committed:
Fix SortableProjectTopicSelector bug on add topic

luucvanderzee committed:
Sort projectAllowedInputTopics by ordering

luucvanderzee committed:
_

luucvanderzee committed:
Fix everything disappearing in SortableProjectTopicList on delete topic

luucvanderzee committed:
Add spinner to SortableProjectTopicList if refetching data

luucvanderzee committed:
Refetch project endpoint on reorder

luucvanderzee committed:
Show correct title while drag/drop

luucvanderzee committed:
Reduce re-renders on drag/drop

luucvanderzee committed:
Fix glitchy drag and drop behavior

luucvanderzee committed:
Making sure that useProjectAllowedInputTopics never returns an invalid ordering

luucvanderzee committed:
Fix linter errors

luucvanderzee committed:
Remove dead code

alexander-cit committed:
Merge branch 'master' into en-27-create-new-projects-topics-functionality

alexander-cit committed:
Merge branch 'en-27-create-new-projects-topics-functionality' into EN-1870-filter-projects-by-topics

luucvanderzee committed:
_

luucvanderzee committed:
Remove console.logs

luucvanderzee committed:
Caching thing solved it seems!

luucvanderzee committed:
Add some comments to document weird streams.ts solution

alexander-cit committed:
[EN-1870] Fetch projects_topics from separate endpoint

It turned out that topics caching doesn't work well on FE,
so we've moved this data from projects#show here.

This reverts commits
39122df9c137dab4d45c350853f58aa7172b9ad6
and
6bf0b77d87bebcfaf7e717bcb1d131ecc919ff5c

luucvanderzee committed:
Merge branch 'EN-1870-filter-projects-by-topics' into CL-339-i3-refactor-terminology-and-endpoint-usage

luucvanderzee committed:
Add listProjectAllowedInputTopics service, remove projectAllowedInputTopicStream and refactor useProjectAllowedInputTopics

luucvanderzee committed:
Update component library

luucvanderzee committed:
Refactor GetProjectAllowedInputTopics

luucvanderzee committed:
Add check for valid ordering in subscription in useProjectAllowedInputTopics and resource

luucvanderzee committed:
Update IProjectData typing

luucvanderzee committed:
Fix CI errors

luucvanderzee committed:
Small typing improvement IdeaForm

luucvanderzee committed:
Try to fix package-lock.json again

luucvanderzee committed:
Remove unnecessary memoization

luucvanderzee committed:
Fix infinite update in useTopics and respond to changes in query params

luucvanderzee committed:
Merge pull request #1535 from CitizenLabDotCo/CL-339-i3-refactor-terminology-and-endpoint-usage

CL-339 i3: Refactor topic terminology and endpoint usage

alexander-cit committed:
Merge branch 'master' into EN-1870-filter-projects-by-topics

luucvanderzee committed:
_

luucvanderzee committed:
Fix project_topics test

luucvanderzee committed:
Skip email_consent again (even though it works perfectly on my machine)

luucvanderzee committed:
Also skipping profile-deletion

alexander-cit committed:
Merge branch 'master' into EN-1870-filter-projects-by-topics

alexander-cit committed:
Merge branch 'master' into EN-1870-filter-projects-by-topics

alexander-cit committed:
Merge pull request #1468 from CitizenLabDotCo/EN-1870-filter-projects-by-topics

[EN-1870] Filter projects by topics
